opentelemetry-appender-tracing/Cargo.toml

```diff
@@ -11,9 +11,10 @@ license = "Apache-2.0"
 rust-version = "1.65"
 
 [dependencies]
-opentelemetry = { version = "0.21", path = "../opentelemetry", features = ["logs"] }
-opentelemetry_sdk = { version = "0.21", path = "../opentelemetry-sdk", features = ["logs"] }
-tracing = {version = "0.1", default-features = false, features = ["std"]}
+opentelemetry = { path = "../opentelemetry", features = ["logs"] }
+opentelemetry_sdk = { path = "../opentelemetry-sdk", features = ["logs"] }
+tracing = { version = "0.1", default-features = false, features = ["std"] }
+tracing-opentelemetry = "0.22"
```

Comment:
this looks odd to depend on tracing-opentelemetry!

Reply:
Definitely. I believe it's necessary since it's where the OtelData is injected into the current span extension and where the `TraceId` and `SpanId` is generated. If there's a better way to retrieve the data, I would be happy to update the PR and remove this dependency.

Reply:
Can we just copy the `OtelData` definition from tracing-opentelemetry, doesn't seems to be a deep-rooted definition. And this will help get rid of dependency.

```rust
#[derive(Debug, Clone)]
pub struct OtelData {
    /// The parent otel `Context` for the current tracing span.
    pub parent_cx: opentelemetry::Context,

    /// The otel span data recorded during the current tracing span.
    pub builder: opentelemetry::trace::SpanBuilder,
}
```

In the future, we do want to have a single subscriber/layer for both traces and logs and can consolidate at that time.

Reply:
I'm not sure this works as the extension type internally uses a `HashMap<TypeId, _>`, and the TypeId of the copied definition would happen to be different. I tried an simpler example [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=09af3bc9eea68f47775794e44deea0bf) and the types are indeed not equal.

Reply:
Thanks for the explanation. You are right, copying `OtelData` definition won't work here. Do you think we can make this dependency optional? And then check for the tracing context only if this dependency is met. This will ensure that the users who are using this crate solely for logging purpose are not paying cost of context lookup when no span is actually created. And this is valid also for users who are directly using the otel tracing API ( though we don't recommend it). 

As a side-note, as of now, when using both`tracing-opentelemetry` and `opentelemetry-appender-tracing`, the logs emitted with tracing macros (trace!, info!, event! etc) are exported both through SpanExporter as span-events, and the LogExporter as logs. This mayn't be desirable. It would be eventually good to integrate both `tracing-opentelemetry` and `opentelemetry-appender-tracing` crates.

Reply:
Yes definitely, I'll see what I can do. I'll push an update ASAP 👍

Reply:
@maximebedard Just wondering if you are able to proceed further, or any help is required on this?

Reply:
Appologies for the delay @lalitb, I was able to get the CLA situation sorted out and pushed a few commits to fix the dependency issues. I've gated the feature with the flag `tracing` for the lack of a better name, if you have any suggestions I'm all ears!

Reply:
Thanks @maximebedard 
Regarding standardizing flags naming, it is still under discussion, but close to finalise as
 - use `otel_unstable` for specs-compliant features that are not yet stable. These are temporary flags and cleaned up regularly.
 - use `<signal>-<feature-name>` for all other features. 

I think this feature falls under the second category. The signal is `logs` here, so we can use `logs_<feature-name>`. Eg `logs_tracing_context` or `logs_propogate_tracing_context`. You can decide on the name acccordingly.

Reply:
What else do we need to complete to get this merged?

Reply:
There are a few merge conflicts, and @maximebedard needs to fix the flag name as above.

Reply:
Seems like this PR has stagnated a bit, anything I can do to help move things forward? @maximebedard I could potentially take over if you don't have the time?

Reply:
Appologies for not getting back to this one, but I don't think it's possible to fix it within the opentelemetry workspace as importing `tracing-opentelemetry` causes a circular dependency. Ultimately, the `OtelData` structure should probably be moved into the opentelemetry-sdk and consumed by the tracing-opentelemetry crate, which would avoid importing tracing-opentelemetry.

I'll close this MR since I can't make anymore progress on it until the `OtelData` structure is moved (see https://github.com/tokio-rs/tracing-opentelemetry/issues/33).

Reply:
So correct me if I'm wrong but this is what we need:
- Create PR here to create OtelData struct
- Create PR in `tracing-opentelemetry` to switch to using that struct
- Recreate this PR without the circular dep

I could begin that process if it would help

Reply:
Thanks @ewoolsey if you could give it a try.

Reply:
> Thanks @ewoolsey if you could give it a try.

@lalitb which crate do you think `OtelData` belongs in?

Reply:
Also, I just pulled this PR and I'm not getting any circular dependencies... Everything is compiling just fine for me.

Reply:
@ewoolsey Can you validate if the span_id and trace_id are populated correctly?

Reply:
Yes it's working correctly. I've got a project depending on this PR and it's functioning as expected.

Reply:
Ok, interesting. I think there won't be any error thrown by the Rust compiler for the circular dependency. Rust strong type system just allows that, but still this is a bad design. One option could be to reopen the PR for review to unblock, and follow up the circular dependency problem with a separate GitHub issue. @maximebedard - if you have any comment?

Reply:
I would love that personally. This is a pretty critical feature for my project and finding a way around it has been really tough.

Reply:
The dependency issue is specific to the workspace, it would not be an issue if you use a released version of the opentelemetry crates and tracing-opentelemetry. We can see the issue in the CI job that failing here https://github.com/open-telemetry/opentelemetry-rust/actions/runs/7105523258/job/19342898991. It's not really possible to push a fix upstream unless we move the `tracing_opentelemetry::OtelData` into the current workspace.

I have a forked version that uses this patch, but I noticed that it doesn't fully solve the problem. The trace_id is only set for the root span (see https://github.com/tokio-rs/tracing-opentelemetry/blob/v0.1.x/src/layer.rs#L871-L874), any child span created would be missing the trace_id and not be correlated correctly.